src/basicfuns.jl

```diff
@@ -119,13 +119,18 @@ end
 
 ## logsumexp
 
-logsumexp{T<:Real}(x::T, y::T) = x > y ? x + log1p(exp(y - x)) : y + log1p(exp(x - y))
+function logsumexp{T<:Real}(x::T, y::T)
+    x == y && abs(x) == Inf && return x
+    x > y ? x + log1p(exp(y - x)) : y + log1p(exp(x - y))
+end
+
 logsumexp(x::Real, y::Real) = logsumexp(promote(x, y)...)
 
 function logsumexp{T<:Real}(x::AbstractArray{T})
-    isempty(x) && return -Inf
+    isempty(x) && return -T(Inf)
     u = maximum(x)
-    s = 0.
+    abs(u) == Inf && return any(isnan(x)) ? T(NaN) : u
```

Comment:
This should be `any(isnan, x)` not `any(isnan(x))`. The first version is `mapreduce`, i.e. applies `isnan` to each element before the `any` reduction`.